domain/model/service/service.go

```diff
@@ -99,3 +163,42 @@ func (s *Service) UpdateCredential(
 
 	return s.st.UpdateCredential(ctx, uuid, id)
 }
+
+// validateAgentVersion is responsible for checking that the agent version that
+// is about to be chosen for a model is a valid to be used.
+//
+// If the agent version is equal to that of the currently running controller
+// then this will be allowed.
+//
+// If the agent version is greater then that of the currently running controller
+// then a [modelerrors.AgentVersionNotSupported] error is returned as
+// we can't run a agent version that is greater then that of a controller.
+//
+// If the agent version is less then that of the current controller we use the
+// toolFinder to make sure that we have tools available for this version. If no
+// tools are available to support the agent version a
+// [modelerrors.AgentVersionNotSupported] error is returned.
+func validateAgentVersion(agentVersion version.Number, toolFinder ToolFinder) error {
+	n := agentVersion.Compare(jujuversion.Current)
+	switch {
+	// agentVersion is greater then that of the current version.
+	case n > 0:
+		return fmt.Errorf(
+			"%w %q cannot be greater then the controller version %q",
+			modelerrors.AgentVersionNotSupported,
+			agentVersion.String(), jujuversion.Current.String(),
+		)
+	// agentVersion is less then that of the current version.
+	case n < 0:
+		err := toolFinder.HasToolsForVersion(agentVersion)
+		if errors.Is(err, errors.NotFound) {
```

Comment:
This ignores an error that isn't NotFound. Using the return signature of `bool,error` as suggested earlier would have made this more obvious.